win32: inject a socket-aware version of CloseHandle() into the CRT

xenu opened the pull request:
_close() on an fd calls CloseHandle(), which is illegal if the fd contains a socket handle.

We previously worked around this by having our own close(), which called closesocket() before calling _close() (https://github.com/Perl/perl5/commit/e601c439adce167078ac7b49550c0418ace86f94).

Amusingly, the author of that solution thought it's just a temporary workaround:
```
  /*
   * close RTL fd while respecting sockets
   * added as temporary measure until PerlIO has real
   * Win32 native layer
   *   -- BKS, 11-11-2000
   */
```
To make it thread-safe, we had to manipulate the internals of file descriptors, which kept changing (https://github.com/Perl/perl5/commit/b47a847f6284f6f98ad7509cf77a4aeb802d8fce).

Unfortunately, the C runtime has been rewritten and it no longer exposes them at all. We had to disable the thread-safety fix in Visual C++ 2015 builds (https://github.com/Perl/perl5/commit/1f664ef5314fb6e438137c44c95cf5ecdbdb5e9b). It also wouldn't work with MinGW configured to use UCRT.

This commit introduces a new solution: we inject a socket-aware version of CloseHandle() into the C runtime library. Hopefully, this will be less fragile.

This also fixes a few issues that the original solution didn't:

- Closing a busy pipe doesn't cause a deadlock (fixes #19963)

- _dup2 properly closes an overwritten socket (fixes #20920)

xenu committed:
win32: initialise winsock unconditionally

The next commit needs it. Also, I don't think this optimisation is worth
the trouble.

leonerd commented on win32/win32.c:
```diff
@@ -5180,6 +5167,168 @@ ansify_path(void)
+            if (strEQ(name, fun_name)) {
+                DWORD old_protect = 0;
+                BOOL succ = VirtualProtect(
+                    &iat->u1.Function, 8, PAGE_READWRITE, &old_protect
```
I don't know the code at all but I'm purely *guessing* that this 8 here is the number of bytes of the range, i.e. expecting to be the size of the pointer. Should that be a `sizeof()` expression? Moreover maybe you want to extract that and the one on line 5211 into a `#define` or something?

mauke replied:
Yeah, that `8` should likely be `sizeof iat->u1.Function`.

xenu replied:
I definitely should've used `sizeof`. Fixed.

xenu committed:
win32: inject a socket-aware version of CloseHandle() into the CRT

_close() on an fd calls CloseHandle(), which is illegal if the fd
contains a socket handle.

We previously worked around this by having our own close(), which called
closesocket() before calling _close()
(e601c439adce167078ac7b49550c0418ace86f94).

Amusingly, the author of that solution thought it's just a temporary
workaround:

  /*
   * close RTL fd while respecting sockets
   * added as temporary measure until PerlIO has real
   * Win32 native layer
   *   -- BKS, 11-11-2000
   */

To make it thread-safe, we had to manipulate the internals of file
descriptors, which kept changing
(b47a847f6284f6f98ad7509cf77a4aeb802d8fce).

Unfortunately, the C runtime has been rewritten and it no longer exposes
them at all. We had to disable the thread-safety fix in Visual C++ 2015
builds (1f664ef5314fb6e438137c44c95cf5ecdbdb5e9b). It also wouldn't work
with MinGW configured to use UCRT.

This commit introduces a new solution: we inject a socket-aware version
of CloseHandle() into the C runtime library. Hopefully, this will be
less fragile.

This also fixes a few issues that the original solution didn't:

  - Closing a busy pipe doesn't cause a deadlock (fixes #19963)

  - _dup2 properly closes an overwritten socket (fixes #20920)
